### ISSUE TITLE
Fiber waiting on ConditionVariable not woken by signal

### ISSUE THREAD

**Issue by brendar** [socketry/async#98]
It appears that a Fiber waiting on `ConditionVariable#wait` is not woken up when another Fiber calls `ConditionVariable#signal`. I encountered this issue while building a prototype with Falcon + Roda + Sequel on Ruby 3.0, where the issue manifests as a `Sequel::PoolTimeout` whenever there are more fibers running than connections available, even after a connection is returned to the pool. The author seems to have been aware of the issue when fiber support was added prior to the release of Ruby 3.0 in https://github.com/jeremyevans/sequel/commit/7c138d86b32cae40e884947167bfe285cbf81d54 but I believe this would be expected to work under Ruby 3.0.

**Reproduction script**
The "waiter" task is only woken after the 5 second timeout has elapsed, not immediately after `signal` is called by the "signaler" task.
```ruby
require 'async'

MUTEX = Mutex.new
CONDITION = ConditionVariable.new
TIMEOUT = 5.0

class Logger
  def self.debug(message)
    task = Async::Task.current
    fiber = Fiber.current
    puts "[#{Time.now}] #{task} on Fiber 0x#{fiber.object_id.to_s(16)}: #{message}"
  end
end

Async do
  Async(annotation: 'waiter') do
    Logger.debug('Starting')

    MUTEX.synchronize do
      Logger.debug('Got the mutex. Waiting for signal.')

      start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      CONDITION.wait(MUTEX, TIMEOUT)
      elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time

      Logger.debug("Wait returned after #{elapsed} seconds")
    end

    Logger.debug('Done')
  end

  Async(annotation: 'signaler') do
    Logger.debug('Starting')

    MUTEX.synchronize do
      Logger.debug('Got the mutex. Signaling')
      CONDITION.signal
    end

    Logger.debug('Done')
  end
end
```

**My environment and output**
```
$ uname -a
Linux toppy 4.15.0-132-generic #136-Ubuntu SMP Tue Jan 12 14:58:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

$ ruby -v
ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-linux]

$ gem list --exact async
*** LOCAL GEMS ***

async (1.28.4)

$ ruby async_condition_variable_test.rb 
[2021-01-27 22:29:16 -0500] #<Async::Task:0x3c waiter (running)> on Fiber 0x50: Starting
[2021-01-27 22:29:16 -0500] #<Async::Task:0x3c waiter (running)> on Fiber 0x50: Got the mutex. Waiting for signal.
[2021-01-27 22:29:16 -0500] #<Async::Task:0x78 signaler (running)> on Fiber 0x8c: Starting
[2021-01-27 22:29:16 -0500] #<Async::Task:0x78 signaler (running)> on Fiber 0x8c: Got the mutex. Signaling
[2021-01-27 22:29:16 -0500] #<Async::Task:0x78 signaler (running)> on Fiber 0x8c: Done
[2021-01-27 22:29:21 -0500] #<Async::Task:0x3c waiter (running)> on Fiber 0x50: Wait returned after 5.004459320000024 seconds
[2021-01-27 22:29:21 -0500] #<Async::Task:0x3c waiter (running)> on Fiber 0x50: Done
```

Please correct me if I've misunderstood the expectations for the fiber scheduler and ConditionVariable. I've just recently started exploring the Socketry projects and the scheduler changes in Ruby 3.0, so it's entirely possible that I've missed something.

Lastly, thanks for all your hard work on the Socketry projects and the fiber scheduler. This is a really exciting new direction for Ruby!

**Comment by ioquatix** [comment]
Thanks for your bug report and the included repro, I'll take a look at what's going on.

**Comment by ioquatix** [comment]
It seems like a bug in `async`.

**Comment by ioquatix** [comment]
Can you please confirm this is fixed? Thanks.

**Comment by brendar** [comment]
Thanks for the quick fix! The change definitely fixes this issue and eliminates the `Sequel::PoolTimeout` in my prototype when I run with 1 Falcon process, 1 pooled database connection, and 2 concurrent requests, but I'm seeing `Sequel::PoolTimeout` errors when I increase to 4 concurrent requests. Not sure if it's related, an unrelated bug, or a starvation issue. I'll investigate and report back.

**Comment by ioquatix** [comment]
Thanks, let me know how you get on, please open a new issue.

**Comment by brendar** [comment]
The remaining timeouts turned out to be starvation. Created new issue https://github.com/socketry/async/issues/99
